### PR TITLE
Updating startup.go for Kubernetes Datastore Driver with bird

### PR DESCRIPTION
**Pull request by heschlie**
This PR coincides with a PR for libcalico-go and projectcalico/confd here:

projectcalico/libcalico-go#354
projectcalico/confd#4

With those two PRs this will allow calico/node to operate with the Kubernetes Datastore Driver and Bird BGP routing enabled in node-mesh mode.

- Updated the shell scripts to detect if running KDD and launch confd accordingly
- startup.go modifications to push all of the required info to KDD

### CODE REVIEW

File: calico_node/startup/startup.go [hunk 1]
```diff
@@ -65,32 +63,23 @@ func main() {
...
-	// If this is a Kubernetes backed datastore then just make sure the
-	// datastore is initialized and then exit.  We don't need to explicitly
+	// If this is a Kubernetes backed datastore then we need to make sure the
+	// datastore is initialized.  We don't need to explicitly
 	// initialize the datastore for non-Kubernetes because the node resource
 	// management will do that for us.
 	if cfg.Spec.DatastoreType == api.Kubernetes {
 		message("Calico is using a Kubernetes datastore")
-		err = client.EnsureInitialized()
+		err := client.EnsureInitialized()
```
Comment (caseydavenport) [inline calico_node/startup/startup.go]: Hm.. This means we're calling EnsureInitialized twice - can we remove this?

Comment (heschlie) [inline calico_node/startup/startup.go]: Is `EnsureInitialized()` called in `watForConnection()`?  I need to double check but IIRC `startup.go` was failing if this was not called, but let me try again and see if this can indeed be removed.

Comment (heschlie) [inline calico_node/startup/startup.go]: If I remove this We end up getting an error related to IPPools:

```
heschlie@Dalaran:vagrant$ kubectl logs -n kube-system calico-node-w1xbv calico-node
Checking datastore connection
time="2017-03-17T21:56:05Z" level=info msg="Loading config from environment" 
time="2017-03-17T21:56:05Z" level=info msg="Datastore is ready" 
Datastore connection verified
Using autodetected IPv4 address on interface eth1: 172.18.18.102/24
No AS number configured on node resource, using global value
time="2017-03-17T21:56:05Z" level=info msg="Initialise BGP data" 
time="2017-03-17T21:56:05Z" level=info msg="Ensure default IPv4 pool is created. IPIP mode: off" 
ERROR: Failed to create default IPv4 IP pool: resource does not exist: IPPool(cidr=10.244.0.0/16)
Terminating
Calico node failed to start
```

But with the `EnsureInitialized()` called it starts successfully:

```
heschlie@Dalaran:vagrant$ kubectl logs -n kube-system calico-node-j1qqp calico-node
Checking datastore connection
time="2017-03-17T22:02:56Z" level=info msg="Loading config from environment" 
Datastore connection verified
Calico is using a Kubernetes datastore
time="2017-03-17T22:02:56Z" level=info msg="Datastore is ready" 
time="2017-03-17T22:02:56Z" level=info msg="Ensuring ThirdPartyResources exist" 
time="2017-03-17T22:02:56Z" level=info msg="Ensuring GlobalConfig ThirdPartyResource exists" 
time="2017-03-17T22:02:56Z" level=info msg="Ensuring IP Pool ThirdPartyResource exists" 
time="2017-03-17T22:02:56Z" level=info msg="ThirdPartyResources exist" 
time="2017-03-17T22:02:56Z" level=info msg="Ensuring ClusterType is set" 
time="2017-03-17T22:02:56Z" level=warning msg="Failed to apply ClusterType: resource does not exist: GlobalConfig(name=ClusterType)" 
time="2017-03-17T22:02:57Z" level=warning msg="Failed to apply ClusterType: resource does not exist: GlobalConfig(name=ClusterType)" 
time="2017-03-17T22:02:58Z" level=warning msg="Failed to apply ClusterType: resource does not exist: GlobalConfig(name=ClusterType)" 
time="2017-03-17T22:02:59Z" level=warning msg="Failed to apply ClusterType: resource does not exist: GlobalConfig(name=ClusterType)" 
time="2017-03-17T22:03:00Z" level=warning msg="Failed to apply ClusterType: resource does not exist: GlobalConfig(name=ClusterType)" 
time="2017-03-17T22:03:01Z" level=warning msg="Failed to apply ClusterType: resource does not exist: GlobalConfig(name=ClusterType)" 
Using autodetected IPv4 address on interface eth1: 172.18.18.101/24
No AS number configured on node resource, using global value
time="2017-03-17T22:03:02Z" level=info msg="ClusterType is set" 
time="2017-03-17T22:03:02Z" level=info msg="Using previously configured cluster GUID" 
time="2017-03-17T22:03:02Z" level=info msg="Kubernetes is initialized as a Calico datastore" 
time="2017-03-17T22:03:02Z" level=info msg="Initialise BGP data" 
time="2017-03-17T22:03:02Z" level=info msg="IPv6 support disabled through environment" 
time="2017-03-17T22:03:02Z" level=info msg="Ensuring ThirdPartyResources exist" 
time="2017-03-17T22:03:02Z" level=info msg="Ensuring GlobalConfig ThirdPartyResource exists" 
time="2017-03-17T22:03:02Z" level=info msg="Ensuring IP Pool ThirdPartyResource exists" 
time="2017-03-17T22:03:02Z" level=info msg="ThirdPartyResources exist" 
time="2017-03-17T22:03:02Z" level=info msg="Ensuring ClusterType is set" 
time="2017-03-17T22:03:02Z" level=info msg="ClusterType is set" 
time="2017-03-17T22:03:02Z" level=info msg="Using previously configured cluster GUID" 
Using node name: 172.18.18.101
time="2017-03-17T22:03:03Z" level=info msg="Loading config from environment" 
Calico node started successfully
snip...
```

I did notice that the `client.nodes.Apply()`, `client.nodes.Create()`, and `client.nodes.Update()` all seem to call `EnsureInitialized()` but `client.nodes.Get()` does not.

Comment (caseydavenport) [inline calico_node/startup/startup.go]: Is it possible to do IP Pools stuff _after_ the node.Apply()?

Comment (heschlie) [inline calico_node/startup/startup.go]: I don't see why not, I'll move it and verify.

Comment (heschlie) [inline calico_node/startup/startup.go]: That works!

File: calico_node/filesystem/etc/rc.local [hunk 2]
```diff
@@ -57,18 +57,24 @@ case "$CALICO_NETWORKING_BACKEND" in
...
+		# confd needs a "-node" arguments for each etcd endpoint.
+		ETCD_ENDPOINTS_CONFD=`echo "-node=$ETCD_NODE" | sed -e 's/,/ -node=/g'`
 
-	confd -confdir=/etc/calico/confd -onetime ${ETCD_ENDPOINTS_CONFD} -no-discover \
-	      -client-key=${ETCD_KEY_FILE} -client-cert=${ETCD_CERT_FILE} \
-	      -client-ca-keys=${ETCD_CA_CERT_FILE} -keep-stage-file >/felix-startup-1.log 2>&1 || true
-	confd -confdir=/etc/calico/confd -onetime ${ETCD_ENDPOINTS_CONFD} -no-discover \
-	      -client-key=${ETCD_KEY_FILE} -client-cert=${ETCD_CERT_FILE} \
-	      -client-ca-keys=${ETCD_CA_CERT_FILE} -keep-stage-file >/felix-startup-2.log 2>&1 || true
+        confd -confdir=/etc/calico/confd -onetime ${ETCD_ENDPOINTS_CONFD} -no-discover \
```
Comment (caseydavenport) [inline calico_node/filesystem/etc/rc.local]: You've got mixed tabs / spaces in this file.

File: calico_node/startup/startup.go [hunk 3]
```diff
@@ -115,10 +87,16 @@ func main() {
 		terminate()
 	}
 
-	// Set other Felix config that is not yet in the node resource.
-	if err := ensureDefaultConfig(client, node); err != nil {
-		fatal("Unable to set global default configuration: %s", err)
-		terminate()
+	// Configure IP Pool configuration.
```
Comment (caseydavenport) [inline calico_node/startup/startup.go]: I know you didn't write it.. but this comment could use some improving.

It's about as useful as this right now:

```
// configureIPPools
configureIPPools()
```

Comment (heschlie) [inline calico_node/startup/startup.go]: I'll take a peek at what happens in there and update the comment

Comment (heschlie) [inline calico_node/startup/startup.go]: Actually do we need a comment here, the name itself is pretty descriptive of what it does and the doctring sheds a bit more light for those curious could we just remove it?  Most of the comments in this section are like that.

Comment (caseydavenport) [inline calico_node/startup/startup.go]: SGTM

**Comment by caseydavenport** [comment]
@heschlie make sure you squash before this gets merged.

**Comment by caseydavenport** [comment]
@heschlie shouldn't there be Makefile changes to use the updated libcalico-go and confd required for this to work?

**Comment by heschlie** [comment]
@caseydavenport Yep, it's coming in my last update, just needed to get a release of confd cut (did yesterday) stay tuned...

**Comment by heschlie** [comment]
@caseydavenport Should be ready for merge, want to give the makefile and glide changes a quick peek?
